File: src/main/java/com/example/dossiermedicale/Entities/TestLaboratoire.java
package com.example.dossiermedicale.Entities;

import jakarta.persistence.*;
import lombok.*;

/**
 * Catalogue des examens disponibles — pas les résultats. Un résultat est une instance d'un test.
 * Les normes (valeurs par âge/sexe) sont dans NormePediatriqueLabo / ConstanteVitale.
 */
@Entity
@Table(name = "test_laboratoire")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class TestLaboratoire {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long idTestLaboratoire;

    @Column(nullable = false, unique = true, length = 50)
    private String codeTest;

    /** Code LOINC standard international (ex: 2160-0). */
    @Column(name = "code_loinc", length = 20)
    private String codeLoinc;

    @Column(nullable = false, length = 255)
    private String nomTest;

    @Column(length = 100)
    private String categorie;

    /** Sang, Urine 24h, Urine spot, Calculé. */
    @Column(name = "type_echantillon", length = 50)
    private String typeEchantillon;

    @Column(length = 50)
    private String unite;

    /** Supprimé du schéma : les normes sont dans NormePediatriqueLabo (âge, sexe). Conservé pour migration. */
    @Deprecated
    @Column(columnDefinition = "TEXT")
    private String valeursNormales;

    @Column(length = 255)
    private String methodeAnalyse;

    private Double prix;

    /** Délai normal de rendu en heures (ex: 4, 24, 48). */
    @Column(name = "delai_rendu_heures")
    private Integer delaiRenduHeures;

    /** Important en pédiatrie. */
    @Column(name = "necessite_jeune", nullable = false)
    @Builder.Default
    private Boolean necessiteJeune = false;
}

File: src/app/services/test-laboratoire.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';

export interface TestLaboratoire {
  idTestLaboratoire?: number;
  codeTest: string;
  codeLoinc?: string;
  nomTest: string;
  categorie?: string;
  typeEchantillon?: string;
  unite?: string;
  /** Déprécié : les normes sont dans NormePediatriqueLabo (âge, sexe). */
  valeursNormales?: string;
  methodeAnalyse?: string;
  prix?: number;
  /** Délai normal de rendu en heures (ex: 4, 24, 48). */
  delaiRenduHeures?: number;
  /** Important en pédiatrie. */
  necessiteJeune?: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class TestLaboratoireService {
  private apiUrl = '/api/tests-laboratoire';

  constructor(private http: HttpClient) {}

  create(dto: TestLaboratoire): Observable<TestLaboratoire> {
    return this.http.post<TestLaboratoire>(this.apiUrl, dto).pipe(
      catchError(err => {
        const msg = err?.error?.message ?? err?.message ?? 'Erreur lors de la création du test';
        return throwError(() => ({ message: msg, error: err?.error }));
      })
    );
  }

  getAll(): Observable<TestLaboratoire[]> {
    return this.http.get<TestLaboratoire[]>(this.apiUrl).pipe(
      catchError(err => throwError(() => err?.message || 'Erreur lors du chargement des tests'))
    );
  }

  getById(id: number): Observable<TestLaboratoire> {
    return this.http.get<TestLaboratoire>(`${this.apiUrl}/${id}`).pipe(
      catchError(err => throwError(() => err?.message || 'Erreur'))
    );
  }
}
